# Stats 21: Homework 1

## Zachary Fann

Acknowledgements: several of these problems are copied from or modified from Think Python by Allen Downey.

I've started the homework file for you. You'll need to fill in the rest with your answers.

My encouragement is to use the keyboard shortcuts as much as possible and use the mouse as little as possible while working the Jupyter Notebook.

After you complete the homework with your answers, in the toolbar, select "Clear All Outputs" and then click "Run All".

Review the document to **make sure all requested output is visible**. You will not get credit for problems where the requested output is not visible, even if the function you coded is correct.

When you are satisfied with the output, click the `...` (to the right of Outline), select "Export", and then Export to HTML.Finally, open the HTML file and "Print as PDF". Submit the PDF to Gradescope.

Submit this ipynb file, complete with your answers and output to Canvas / Bruin Learn.

**Again, you must make sure all requested output is visible to receive full credit.**

# Task 1

Create an account on GitHub. 

Change your profile picture. Ideally, use photo of yourself that would be appropriate for a resume. If you are not comfortable with the idea of using a photo of yourself, use any other image that is suitable for a workplace environment.

Follow the instructions provided in class to fork the class repository to your GitHub.

Create another repository with at least two text files in it on GitHub (other than the forked class notes repository). Make at least two additional commits to the reposiotry and push them to GitHub.

Provide a link to both repositories here.

You will also need to submit the link to your own repository (not the forked one) to Canvas / Bruin Learn.

## Your Answer:

- Link to your forked repository: https://github.com/zacharyfann/26-spring-stats-21.git
- Link to your own repository: https://github.com/zacharyfann/Stats-21-Repo.git

# Problem 2

An important part of programming is learning to interpret error messages and understanding what correction needs to be made.

Read and familiarize yourself with the following error messages.

Explain the error. Then duplicate each cell and correct the error. The first problem has been done for you as an example.

In [19]:
# A
print("Hello World"

SyntaxError: incomplete input (1427017300.py, line 2)

Answer: The `print()` function is missing the closing parenthesis. This results in an unexpected EOF error.

In [20]:
# corrected:
print("Hello World")

Hello World


In [21]:
# B
print("Hello")
    print("Goodbye")

IndentationError: unexpected indent (66443029.py, line 3)

The print line for Goodbye is on a different indentation as the one above resulting in an indentation error

In [22]:
# B
print("Hello")
print("Goodbye")

Hello
Goodbye


In [23]:
# C
x = 10
if x > 8
    print("x is greater than 8")

SyntaxError: expected ':' (4211757722.py, line 3)

The if-statement is missing a colon which results in a syntax error

In [24]:
# C corrected:
x = 10
if x > 8:
    print("x is greater than 8")

x is greater than 8


In [25]:
# D
if x = 10:
    print("x is equal to 10")

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (2686327292.py, line 2)

The if statement is not declaring a variable, it needs to be used as a comparison operator which requires two "=="

In [26]:
# D corrected:
if x == 10:
    print("x is equal to 10")

x is equal to 10


In [27]:
# E
x = 5
if x == 5:
print("x is five")

IndentationError: expected an indented block after 'if' statement on line 3 (2346592504.py, line 4)

there is an indentation error in the if statement because the contents within the if statement have to be on a seperate indentation than when it's initialized

In [28]:
# E corrected:
x = 5
if x == 5:
    print("x is five")

x is five


In [29]:
# F
l = [1, 2, 50, 10]
l = sort(l)

NameError: name 'sort' is not defined

sort is not a function in base python which means you need to use the sorted function

In [30]:
# F corrected:
l = [1, 2, 50, 10]
l = sorted(l)
print(l)

[1, 2, 10, 50]


# Problem 3

Use Python as a calculator. Enter the appropriate calculation in a cell and be sure the output value is visible.

A. How many seconds are there in 42 minutes 42 seconds?

In [31]:
seconds = 42*60 +42
print(seconds)

2562


B. There are 1.61 kilometers in a mile. How many miles are there in 10 kilometers?

In [32]:
miles = 10/1.61 
print(miles)

6.211180124223602


C. If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average 1-mile pace (time to complete 1 mile in minutes and seconds)? What is your average speed in
miles per hour?

In [33]:
print((seconds/miles) //60)

print((seconds/miles) % 60)

hours = (42 + (42/60)) / 60
print(hours)
print(miles / hours)



6.0
52.48200000000003
0.7116666666666667
8.727653570337614


The average 1-mile pace was 6 minutes and 52.48 seconds. 

The average speed was 8.728 mph.

# Problem 4
Write functions for the following problems.

A. The volume of a sphere with radius r is $$V = \frac{4}{3}\pi r^3$$

Write a function `sphere_volume(r)` that will accept a radius as an argument and return the volume.

- Use the function to find the volume of a sphere with radius 5.
- Use the function to find the volume of a sphere with radius 15.

In [34]:
import math

def sphere_volume(r):
    return (4/3) * math.pi * r**3

print(sphere_volume(5))
print(sphere_volume(15))

523.5987755982989
14137.166941154068


B. Suppose the cover price of a book is \\$24.95, but bookstores get a 40\% discount. Shipping costs \\$3 for the first copy and 75 cents for each additional copy.

Write a function `wholesale_cost(books)` that accepts an argument for the number of books and will return the total cost of the books plus shipping.

- Use the function to find the total wholesale cost for 60 copies.
- Use the function to find the total wholesale cost for 10 copies.

In [35]:
def wholesale_cost(books):
    if books > 1:
        return 24.95*books*0.6 + 0.75*books + 3
    else:
        return 24.95*books*0.6 + 3
    
print(wholesale_cost(60))
print(wholesale_cost(10)) 

946.1999999999999
160.2


C. A person runs several miles. The first and last miles are run at an 'easy' pace. Other than the first and last miles, the other miles are at a faster pace.

Write a function `run_time(miles, warm_pace, fast_pace)` to calculate the time the runner will take. The function accepts three input arguments: how many miles the runner travels (minimum value is 2), the warm-up and cool-down pace, the fast pace. The function will print the time in the format minutes:seconds, and will return a tuple of values: (minutes, seconds)

Use the function to find the time to run a total of 5 miles. The warm-up pace is 8:15 per mile. The speed pace is 7:12 per mile.

Call the function using: `run_time(miles = 5, warm_pace = 495, fast_pace = 432)`

In [36]:
def run_time(miles, warm_pace, fast_pace):
    warm_time = 2*warm_pace
    fast_time = (miles -2) * fast_pace
    total_time = warm_time + fast_time
    minutes = total_time // 60
    seconds = total_time % 60
    return minutes, seconds

print(run_time(5,495,432))


(38, 6)


Another important skill is to be able to read documentation.

Read the documentation for the function str.split() at https://docs.python.org/3/library/stdtypes.html#str.split

Adjust the function so that the call can be made with minutes and seconds:

`run_time(miles = 5, warm_pace = "8:15", fast_pace = "7:12")`

In [37]:
def run_time(miles, warm_pace, fast_pace):
    warm_time_list = str.split(warm_pace, ':')
    warm_time = (int(warm_time_list[0]) * 60 + int(warm_time_list[1])) * 2
    fast_time_list = str.split(fast_pace, ':')
    fast_time = (int(fast_time_list[0]) * 60 + int(fast_time_list[1])) * (miles - 2)
    total_time = warm_time + fast_time
    minutes = total_time // 60
    seconds = total_time % 60
    return minutes, seconds

print(run_time(5,"8:15","7:12"))


(38, 6)


# Problem 5

Use `import math` to gain access to the math library.

Create a function `polar(real, imaginary)` that will return the polar coordinates of a complex number.

The input arguments are the real and imaginary components of a complex number. The function will return a tuple of values: the value of the radius `r` and the angle `theta`. 

For a refresher, see: https://ptolemy.berkeley.edu/eecs20/sidebars/complex/polar.html

Show the results for the following complex numbers:

- 1 + i
- -2 - 3i
- 4 + 2i

In [38]:
import math

def polar(real, imaginary):
    imaginary_new = imaginary.replace('i', '')
    imaginary_value = int(imaginary_new) if imaginary_new != '' else 0
    return (math.hypot(real, imaginary_value), math.atan2(imaginary_value, real))

print(polar(1, "2i"))
print(polar(-2, "-3i"))
print(polar(4, "2i")    )

(2.23606797749979, 1.1071487177940904)
(3.605551275463989, -2.158798930342464)
(4.47213595499958, 0.4636476090008061)


# Problem 6

Define a function called `insert_into(listname, index, iterable)`. It will accept three arguments, a currently existing list, an index, and another list/tuple that will be inserted at the index position.

Python's built-in function, `list.insert()` can only insert one object.

In [39]:
# write your code here
def insert_into(listname, index, iterable):
    for i, element in enumerate(iterable):
        listname.insert(index + i, element)
    print(listname)


In [40]:
# do not modify. We will check this result for grading
l = [0,'a','b','c',4,5,6]
i = ['hello', 'there']
insert_into(l, 3, i)

[0, 'a', 'b', 'hello', 'there', 'c', 4, 5, 6]


# Problem 7

Define a function called `first_equals_last(listname)`

It will accept a list as an argument. It will return `True` if the first and last elements are equal and the if the list has a length greater than 1. It will return False for all other cases.

In [41]:
# write your function here
def first_equals_last(listname):
    if len(listname) == 0:
        return False
    return listname[0] == listname[-1]

In [42]:
# do not modify. We will check this result for grading
a = [1,2,3]
first_equals_last(a)

False

In [43]:
# do not modify. We will check this result for grading
b = ['hello','goodbye','hello']
first_equals_last(b)

True

In [44]:
# do not modify. We will check this result for grading
c  = [1,2,3,'1']
first_equals_last(c)

False

In [45]:
# do not modify. We will check this result for grading
d = [[1,2],[3,2],[1,2]]
first_equals_last(d)

True